# Getting Started

## Building xAtlas

xAtlas requires the following software components to build:

- [CMake](https://cmake.org/) (version 3.0 or higher required)
- [HTSlib](https://github.com/samtools/htslib) (version 1.6 or higher
recommended)
- libpthread (included with most Unix-like systems)

xAtlas can be built using CMake by running the following commands from the base
directory:

    $ mkdir build
    $ cd build
    $ cmake ..
    $ make

If HTSlib is installed in a location outside of CMake's library search path,
the HTSlib install prefix can be set for the CMake command by defining the
CMake variable `HTSLIB_PREFIX`:

    $ cmake -DHTSLIB_PREFIX=/path/to/htslib-1.x ..

See the official [CMake documentation](https://cmake.org/cmake/help/latest/)
for more options to control build and installation settings.

## Input files

For use with xAtlas, reference genomes should be in FASTA format and have a
FASTA index file (`.fai`), which can be generated by `samtools faidx`. Input
sample alignments should be in either BAM or CRAM format and have an index file
(`.bai`, `.crai`, or `.csi`), which can be generated by `samtools index`. The
examples commands below use files that can be found at the following URLs:

    ## GRCh38 Human reference genome in FASTA format
    http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa
    http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa.fai
    ## 1000 Genomes Project NA12878/HG001 alignment at 30x coverage in CRAM format
    http://ftp.sra.ebi.ac.uk/vol1/run/ERR323/ERR3239334/NA12878.final.cram
    http://ftp.sra.ebi.ac.uk/vol1/run/ERR323/ERR3239334/NA12878.final.cram.crai
    ## NIST Genome in a Bottle v4.2.1 benchmark small variants BED regions for NA12878/HG001
    https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/NA12878_HG001/NISTv4.2.1/GRCh38/HG001_GRCh38_1_22_v4.2.1_benchmark.bed

## Example usage

A full list of command line options can be shown by running `xatlas --help`.

For basic usage with only required arguments, xAtlas requires the input
alignment file, the reference genome file, the sample name to be populated in
the VCF header, and the output filename prefix. This command will run SNP and
indel variant calling on the CRAM alignment and create two VCF files as output,
a SNP VCF named `NA12878_xatlas_snp.vcf` and an indel VCF named
`NA12878_xatlas_indel.vcf`:

    $ xatlas \
        --in NA12878.final.cram \
        --ref GRCh38_full_analysis_set_plus_decoy_hla.fa \
        --sample-name NA12878 \
        --prefix NA12878_xatlas

To limit variant calling output to specific genomic regions, a BED file can be
supplied to the `--capture-bed` option, which will limit variant calls to only
those within regions defined in the BED file.

    $ xatlas \
        --in NA12878.final.cram \
        --ref GRCh38_full_analysis_set_plus_decoy_hla.fa \
        --sample-name NA12878 \
        --prefix NA12878_xatlas \
        --capture-bed HG001_GRCh38_1_22_v4.2.1_benchmark.bed

To enable gVCF-formatted output, include the `--gvcf` option in the xAtlas
command. It is recommended to use this option together with the `--bgzf`
option, which will write the output VCFs using BGZF compression (`.vcf.gz`).

    $ xatlas \
        --in NA12878.final.cram \
        --ref GRCh38_full_analysis_set_plus_decoy_hla.fa \
        --sample-name NA12878 \
        --prefix NA12878_xatlas \
        --gvcf \
        --bgzf

Multithreaded operation can be enabled by `--num-hts-threads`, which enables
setting an additional number of HTSlib BAM/CRAM decompression threads, or by
`--multithread`, which splits the the processes for reading the input
alignment, SNP variant calling, and indel variant calling into separate
threads. These two options can be used separately or in combination:

    $ xatlas \
        --in NA12878.final.cram \
        --ref GRCh38_full_analysis_set_plus_decoy_hla.fa \
        --sample-name NA12878 \
        --prefix NA12878_xatlas \
        --num-hts-threads 4 \
        --multithread
